keygen/src/ssl.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2018 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <openssl/rand.h>
+#include <openssl/ecdsa.h>
+#include <openssl/x509.h>
+
+#include "ssl.h"
+
+/**
+ * gen_EC() - Generate an Elliptic Curve key (NID_X9_62_prime256v1)
+ *
+ * Return: A pointer to a new EC_KEY instance or NULL on failure
+ */
+EC_KEY * gen_EC()
+{
+    EC_KEY * ec_key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+
+    if (ec_key && !EC_KEY_generate_key(ec_key))
+    {
+        EC_KEY_free(ec_key);
+        ec_key = NULL;
+    }
+
+    return ec_key;
+}
+
+/**
+ * gen_cert() - Generate an X509 cert with a public key
+ * @pkey: The public key to certify as an EVP_PKEY*
+ * @cn: The IPv6 address (as a string) to assign to CN
+ *
+ * Return: A pointer to a new X509 certificate
+ */
+X509 *gen_cert(EVP_PKEY *pkey, char *cn)
+{
+    X509 *x509 = X509_new();
```

Comment:
As with the `EC_KEY`, how does this get `free`'d later?

Reply:
I wasn't managing that. I added a destroy function but it's a simple call to `X509_free`